pkg/agent/controller/networkpolicy/fqdn.go

```diff
@@ -746,15 +747,41 @@ func (f *fqdnController) HandlePacketIn(pktIn *ofctrl.PacketIn) error {
 func (f *fqdnController) handlePacketIn(pktIn *ofctrl.PacketIn) error {
 	klog.V(4).InfoS("Received a packetIn for DNS response")
 	waitCh := make(chan error, 1)
-	handleDNSData := func(dnsData []byte) {
+	handleUDP := func(udp *protocol.UDP) {
 		dnsMsg := dns.Msg{}
-		if err := dnsMsg.Unpack(dnsData); err != nil {
+		if err := dnsMsg.Unpack(udp.Data); err != nil {
 			// A non-DNS response packet or a fragmented DNS response is received. Forward it to the Pod.
 			waitCh <- nil
 			return
 		}
 		f.onDNSResponseMsg(&dnsMsg, time.Now(), waitCh)
 	}
+	handleTCP := func(tcpPkt *protocol.TCP) {
+		dnsData, dataLength, err := binding.GetTCPDNSData(tcpPkt)
+		if err != nil {
+			// The packet doesn't contain a valid DNS length field and data. Forward it to the Pod.
+			klog.V(4).InfoS("Unable to get DNS data from the packet, skipping it", "err", err)
+			waitCh <- nil
+			return
+		}
+		dnsMsg := dns.Msg{}
+		if dataLength > len(dnsData) {
+			// This is likely the first packet containing the length field and partial message.
+			klog.InfoS("Received a fragmented DNS response, partially unpacking it", "lengthField", dataLength, "actualLength", len(dnsData))
+			if err := dnsutil.UnpackDNSMsgPartially(dnsData, &dnsMsg); err != nil {
+				klog.InfoS("Unable to unpack the DNS response partially, skipping it", "err", err)
+				waitCh <- nil
+				return
+			}
+		} else if err := dnsMsg.Unpack(dnsData); err != nil {
+			// This is likely a non-DNS response packet or a non-first-DNS response packet containing partial message.
+			// Set verbose level to 2 as normally we are not interested in it.
```

Comment:
nit: could you mention in the comments as well that usually the first fragment already contains host <-> IP information that we are interested in?

Reply:
done